src/cli.js

```diff
@@ -65,71 +69,86 @@ function handleError (error) {
   process.emit('SIGINT')
 }
 
-function parseDocuments (data, primary) {
-  if (primary && !JSON.parse(data)[primary]) {
-    throw new Error(`Is the primary key correct?\n\n${primary}`)
-  }
-
-  try {
-    return primary ? JSON.parse(data)[primary] : JSON.parse(data)
-  } catch (err) {
-    try {
-      return data
-        .toString()
-        .trim()
-        .split('\n')
-        .map(JSON.parse)
-    } catch (error) {
-      throw new Error(`Is the input valid JSON/NDJSON?\n\n${error}`)
-    }
-  }
-}
-
-function check ({ query, file, url, stdIn }) {
+function check ({ query, inputFormat, outputFormat }) {
   if (!query) {
     throw Error(
       chalk.yellow(
         'You must add a query. To learn more, run\n\n  $ groq --help'
       )
     )
   }
-  if (!file && !url && !stdIn) {
-    throw Error(
-      chalk.yellow(
-        'There’s no data to query. To learn more, run\n\n  $ groq --help'
-      )
-    )
+
+  if (!/^(json|ndjson|null)$/.test(inputFormat)) {
+    throw Error(chalk.yellow(`Unknown input format: ${inputFormat}`))
+  }
+
+  if (!/^(json|ndjson|pretty)$/.test(outputFormat)) {
+    throw Error(chalk.yellow(`Unknown output format: ${outputFormat}`))
   }
+
   return true
 }
 
-async function parseQuery () {
+async function* runQuery() {
   const { flags, input } = cli
-  const { file, url, primary, pretty } = flags
+  const { input: inputFormat, pretty } = flags
+  let { output: outputFormat } = flags
+
   const query = input[0]
-  const stdIn = await getStdin()
-  check({ query, file, url, stdIn })
-  let documents = []
-  if (file) {
-    const fileContent = await readFileStream(file).catch(handleError)
-    documents = await parseDocuments(fileContent, primary)
-  } else if (url) {
-    const urlContent = await fromUrl(url)
-    documents = await parseDocuments(urlContent, primary)
-  } else if (stdIn) {
-    documents = await parseDocuments(stdIn, primary)
+  if (pretty) {
+    outputFormat = 'pretty'
   }
 
+  check({ query, inputFormat, outputFormat })
+
   const tree = parse(query)
-  const result = await evaluate(tree, { documents }).get()
 
-  if (pretty) {
-    return colorizeJson(result)
+  let documents
+  let root
+
+  switch (inputFormat) {
+    case 'json':
+      root = JSON.parse(await getStdin())
+      break
+    case 'ndjson':
+      documents = new S2A(process.stdin.pipe(ndjson()))
+      break
+    default:
+      // do nothing
+  }
+
+  const result = await evaluate(tree, { documents, root })
+
+  switch (outputFormat) {
```

Comment:
I prefer object literals to switch statements because I find them less easy to break and better scoped. But as long as this code is short and sweet I guess it's OK.